src/core/lib/event_engine/windows/windows_engine.cc

```diff
@@ -43,20 +45,113 @@
 #include "src/core/lib/gprpp/crash.h"
+#include "src/core/lib/gprpp/dump_args.h"
 #include "src/core/lib/gprpp/sync.h"
 #include "src/core/lib/gprpp/time.h"
 #include "src/core/lib/iomgr/error.h"
 
 namespace grpc_event_engine {
 namespace experimental {
 
-namespace {
-EventEngine::OnConnectCallback CreateCrashingOnConnectCallback() {
-  return [](absl::StatusOr<std::unique_ptr<EventEngine::Endpoint>>) {
-    grpc_core::Crash("Internal Error: OnConnect callback called when unset");
-  };
+std::ostream& operator<<(
+    std::ostream& out,
+    const WindowsEventEngine::ConnectionState& connection_state) {
+  out << "ConnectionState::" << &connection_state
+      << ": connection_state.address="
+      << ResolvedAddressToURI(connection_state.address_) << ","
+      << GRPC_DUMP_ARGS(connection_state.has_run_,
+                        connection_state.connection_handle_,
+                        connection_state.timer_handle_);
+  return out;
 }
-}  // namespace
+
+// ---- ConnectionState ----
+
+WindowsEventEngine::ConnectionState::ConnectionState(
+    std::shared_ptr<WindowsEventEngine> engine,
+    std::unique_ptr<WinSocket> socket, EventEngine::ResolvedAddress address,
+    MemoryAllocator allocator,
+    EventEngine::OnConnectCallback on_connect_user_callback)
+    : engine_(std::move(engine)),
+      socket_(std::move(socket)),
+      address_(address),
+      allocator_(std::move(allocator)),
+      on_connect_user_callback_(std::move(on_connect_user_callback)) {
+  CHECK(socket_ != nullptr);
+  connection_handle_ = ConnectionHandle{reinterpret_cast<intptr_t>(this),
+                                        engine_->aba_token_.fetch_add(1)};
+}
+
+void WindowsEventEngine::ConnectionState::Start(Duration timeout) {
+  on_connected_cb_ =
+      std::make_unique<OnConnectedCallback>(engine_.get(), shared_from_this());
+  socket_->NotifyOnWrite(on_connected_cb_.get());
+  deadline_timer_cb_ = std::make_unique<DeadlineTimerCallback>(
+      engine_.get(), shared_from_this());
+  timer_handle_ = engine_->RunAfter(timeout, deadline_timer_cb_.get());
+}
+
+EventEngine::OnConnectCallback
+WindowsEventEngine::ConnectionState::TakeCallback() {
+  grpc_core::MutexLock lock(&mu_);
+  return std::exchange(on_connect_user_callback_, nullptr);
+}
+
+std::unique_ptr<WindowsEndpoint>
+WindowsEventEngine::ConnectionState::FinishConnectingAndMakeEndpoint(
+    ThreadPool* thread_pool) {
+  ChannelArgsEndpointConfig cfg;
+  return std::make_unique<WindowsEndpoint>(address_, std::move(socket_),
+                                           std::move(allocator_), cfg,
+                                           thread_pool, engine_);
+}
+
+void WindowsEventEngine::ConnectionState::AbortOnConnect() {
+  on_connected_cb_.reset();
+}
+
+void WindowsEventEngine::ConnectionState::AbortDeadlineTimer() {
+  deadline_timer_cb_.reset();
+}
+
+void WindowsEventEngine::ConnectionState::OnConnectedCallback::Run() {
```

Comment:
Nit: Probably code between WindowsEventEngine::ConnectionState::OnConnectedCallback::Run and WindowsEventEngine::ConnectionState::DeadlineTimerCallback::Run can be shared.

Reply:
To eliminate 10 SLOC duplication, options include: introducing a shared base class with protected shared state, adding friend classes/functions to access private members of the ConnectionState, or putting the engine callback invocation in a lambda. I'm not a fan of any, I think this finite (2-count) duplication is best.